src/codegen/debug.rs

```diff
@@ -0,0 +1,518 @@
+use std::{cell::RefCell, collections::HashMap, ops::Range};
+
+use inkwell::{
+    context::ContextRef,
+    debug_info::{
+        AsDIScope, DIBasicType, DICompileUnit, DICompositeType, DIDerivedType, DIFlags,
+        DIFlagsConstants, DIType, DWARFEmissionKind, DebugInfoBuilder,
+    },
+    module::Module,
+    values::{BasicMetadataValueEnum, GlobalValue},
+};
+
+use crate::{
+    ast::SourceRange,
+    datalayout::Byte,
+    diagnostics::Diagnostic,
+    index::Index,
+    typesystem::{DataType, DataTypeInformation, StringEncoding, BOOL_SIZE, CHAR_TYPE, WCHAR_TYPE},
+    DebugLevel, OptimizationLevel,
+};
+
+#[derive(PartialEq, Eq)]
+#[allow(non_camel_case_types)]
+enum DebugEncoding {
```

Comment:
maybe we can find a better name for that enum - i'm not 100% what it does

what does DW_ATE stand for?

Reply:
DW = Dwarf
AT = Attribute
E = Encoding, hence the name.. The parameter this is passed as is also called encoding.

Reply:
The encoding here is usually only used for Basic Types, which I updated in the comment